File: cpp/src/ymlconfig-pp.cpp
#include <regex>
#include <fstream>
using namespace std;

#include <experimental/filesystem>
namespace fs = std::experimental::filesystem;

#include <kvan/ymlconfig-pp.h>
#include <kvan/string-utils.h>

YMLConfigPP::YMLConfigPP(const vector<string>& pp) {
  for (auto& p: pp) {
    pp_pathes.push_back(evaluate_dollar_var_expr(p));
  }
}

pair<string, bool> YMLConfigPP::find_yml_file(const string& yml_fn)
{
  for (auto pp_path: pp_pathes) {
    string yml_pn;
    yml_pn = pp_path;
    yml_pn += "/";
    yml_pn += yml_fn;
    yml_pn = evaluate_dollar_var_expr(yml_pn);
    if (fs::exists(yml_pn)) {
      return make_pair(yml_pn, true);
    }
  }
  return make_pair("", false);
}

void YMLConfigPP::process_include(const string& include_line)
{
  regex re(R"D(!include\s+\<([\w/\.\-]+)\>)D");

  smatch m;
  regex_search(include_line, m, re);
  if (!(m.ready() && m.size() == 2)) {
    throw runtime_error(__func__);
  }

  auto include_file = find_yml_file(m[1]);
  if (include_file.second == false) {
    ostringstream mm; mm << "YMLConfigPP::process_include: can't find " << m[1].str();
    throw runtime_error(mm.str());
  }
  
  if (auto it = this->include_files.find(include_file.first);
      it == this->include_files.end()) {
    this->include_files.insert(include_file.first);
    this->run_pp(include_file.first);
  }
}

void YMLConfigPP::run_pp(const string& yml_fn)
{
  ifstream yml_fn_in(yml_fn);
  if (!yml_fn_in) {
    ostringstream m; m << "YMLConfigPP:: can't read file " << yml_fn;
    throw runtime_error(m.str());
  }

  out << "# file " << yml_fn << endl;
  string l;
  while (getline(yml_fn_in, l)) {
    if (l.find("!include") != string::npos) {
      out << "# " << l << endl;
      process_include(l);
    } else {
      out << l << endl;
    }
  }

  out << "#" << endl;
  out << "# end of file " << yml_fn << endl;
}

File: python/KVAN/ymlconfig_pp.py
from io import StringIO
from .dvexpansion import *

class YMLConfigPP:
    def __init__(self, pathes):
        self.out_fd = StringIO()
        self.include_files = set()
        self.pp_pathes = []
        for p in pathes:
            self.pp_pathes.append(evaluate_dollar_var_expr(p))

    def find_yml_file(self, yml):
        ret = None
        for pp_path in self.pp_pathes:
            print("pp_path: ", pp_path)
            yml_fn = os.path.join(pp_path, yml)
            if os.path.exists(yml_fn):
                ret = yml_fn
                break
        return ret

    def get_pp_content(self):
        return self.out_fd.getvalue()
    
    def run_pp(self, yml_fn):
        fd = open(yml_fn, "r")
        
        self.out_fd.write(f"# {yml_fn}\n")
        while 1:
            l = fd.readline()
            if not l:
                break
            if l.find("!include") != -1:
                self.out_fd.write(f"# {l}")
                self.process_include(l)
            else:
                self.out_fd.write(l)
                
        self.out_fd.write("\n#\n")
        self.out_fd.write(f"# end of file {yml_fn}\n")

    def process_include(self, include_line):
        include_re = r"!include\s+<([\w+/\.\-]+)>"

        m = re.match(include_re, include_line)
        if m == None or len(m.groups()) != 1:
            raise Exception(f"YMLConfigPP::process_include: malformed line {include_line}")

        include_file = self.find_yml_file(m.groups()[0])
        if include_file == None:
            raise Exception(f"YMLConfigPP::process_include: can't resolve {include_line}")

        if not include_file in self.include_files:
            self.include_files.add(include_file)
            self.run_pp(include_file)
        
